### PR TITLE
add aws-sqs to registry

### LINKED ISSUE #307: Add aws-sqs components
# Description
Add an aws-sqs component for creating aws-sqs.
## Inputs
- QueueName 
- DelaySeconds
- MaximumMessageSize
- MessageRetentionPeriod
- ReceiveMessageWaitTimeSeconds
- VisibilityTimeout
- FifoQueue
- ContentBasedDeduplication

## Outputs
- QueueUrl

## Requirements
 

- [x] tests
  

- [x] documentation

  

- [ ] examples

### PR DESCRIPTION
**Pull request by dilantha111**
## What has been implemented?

Closes #307

## Steps to verify

* Create component 

 ### serverless.yml
```
type: my-application
 components:
  myGreeter:
    type: aws-sqs
    inputs:
      queueName: myFirstQueue3
      DelaySeconds: '21'
      MaximumMessageSize: '1024'
```
## Todos:

* [x] Write tests
* [x] Write / update documentation
* [x] Run Prettier

### CODE REVIEW

**Comment by dilantha111** [comment]
Hi @brianneisler  I added some unit tests. Please have a look. But for some reason, npm run docs doesn't generate the docs for aws-sqs. Is there any configs to be added before running "npm run docs" ?

**Comment by laardee** [comment]
@dilantha111 You need to add `example` into `inputTypes` and `outputTypes` to create docs using `npm run docs` e.g.

```yaml
inputTypes:
  QueueName:
    type: string
    displayName: AWS SQS name
    description: The name of the SQS to create
    example: myqueue # <-- this
  DelaySeconds:
    type: string
```

I would also change the inputTypes to start with lowercase, `QueueName` -> `queueName`, as all the other components use that naming convention.

**Comment by dilantha111** [comment]
Hi @laardee  I tried running "npm run docs" after adding example field. But result is still the same.

**Comment by laardee** [comment]
@dilantha111 oh yes, I see, you are missing the whole README.md file. You need to add that to the component directory with following content

```
<!-- AUTO-GENERATED-CONTENT:START (COMPONENT_HEADER) -->
<!-- AUTO-GENERATED-CONTENT:END -->
<!-- AUTO-GENERATED-CONTENT:START (TOC) -->
<!-- AUTO-GENERATED-CONTENT:END -->
<!-- AUTO-GENERATED-CONTENT:START (COMPONENT_INPUT_TYPES) -->
<!-- AUTO-GENERATED-CONTENT:END -->
<!-- AUTO-GENERATED-CONTENT:START (COMPONENT_OUTPUT_TYPES) -->
<!-- AUTO-GENERATED-CONTENT:END -->
<!-- AUTO-GENERATED-CONTENT:START (COMPONENT_EXAMPLES) -->
<!-- AUTO-GENERATED-CONTENT:END -->
```

The `npm run docs` command then fills in those blocks.

File: registry/aws-sqs/src/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    Attributes: attributes
+  }
+
+  return SQS.createQueue(params)
+    .promise()
+    .then((res) => res.QueueUrl)
+}
+
+const deleteQueue = async (state) => {
+  return SQS.deleteQueue(state.queueUrl).promise()
```
Comment (laardee) [inline registry/aws-sqs/src/index.js]: If you delete the queue manually in the AWS web console and then try to remove it with SDK this probably fails. You should catch and ignore the "not found" (or similar) error here. 

Like this https://github.com/serverless/components/blob/master/registry/aws-lambda/src/index.js#L143

Create a test case for that scenario also.

File: registry/aws-sqs/src/index.test.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    mocks,
+    SQS: jest.fn().mockImplementation(() => SQS)
+  }
+})
+
+afterEach(() => {
+  AWS.mocks.createQueueMock.mockClear()
+  AWS.mocks.deleteQueueMock.mockClear()
+})
+
```
Comment (laardee) [inline registry/aws-sqs/src/index.test.js]: Restore all mocks after the test
```js
afterAll(() => {
  jest.restoreAllMocks()
})
```

File: registry/aws-sqs/serverless.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+version: 0.2.0
+core: 0.2.x
+
+description: 'Provision AWS SQS components'
+license: Apache-2.0
+author: 'Serverless, Inc. <hello@serverless.com> (https://serverless.com)'
+repository: 'github:serverless/components'
+
+inputTypes:
+  queueName:
```
Comment (laardee) [inline registry/aws-sqs/serverless.yml]: QueueName is required so you can add `required: true` attribute to it

File: registry/aws-sqs/src/index.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
+const AWS = require('aws-sdk')
+const SQS = new AWS.SQS({ region: process.env.AWS_DEFAULT_REGION || 'us-east-1' })
+const _ = require('lodash')
```
[the remark this answers is not in the document]
Comment (dilantha111) [reply registry/aws-sqs/src/index.js]: Added ramda and removed lodash

File: registry/aws-sqs/src/index.js (hunk 1 above)
Comment (dilantha111) [reply registry/aws-sqs/src/index.js]: Added error handling for the deleteQueue

File: registry/aws-sqs/serverless.yml (hunk 3 above)
Comment (dilantha111) [reply registry/aws-sqs/serverless.yml]: added required: true for queueName

**Comment by dilantha111** [comment]
Hi @laardee and @brianneisler  hope I covered the requested changes. Thank you so much for the help.

**Comment by brianneisler** [comment]
@laardee feel free to merge this after you've given your sign off.

File: registry/aws-sqs/src/index.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  return SQS.createQueue(params)
+    .promise()
+    .then((res) => res.QueueUrl)
+}
+
+const deleteQueue = async (state) => {
+  try {
+    await SQS.deleteQueue(state.queueUrl).promise()
```
Comment (laardee) [inline registry/aws-sqs/src/index.js]: The queue url should be passed to the `deleteQueue` function as an object

```js
await SQS.deleteQueue({ QueueUrl: state.queueUrl }).promise()
```

File: registry/aws-sqs/src/index.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  } catch (error) {
+    if (!error.message.includes('The specified queue does not exist')) {
+      throw new Error(error)
+    }
+  }
+  return null
+}
+
+const capitalizeKeys = (obj) => {
```
Comment (laardee) [inline registry/aws-sqs/src/index.js]: I would rather use reduce in this, something like
```js
const capitalizeKeys = (obj) =>
  reduce((result, key) =>
    merge(result, { [capitalizeString(key)]: obj[key] }), {}, keys(obj))
```
The merge and reduce are from the ramda library in the example.

But this is a matter of opinion, I find reduce to be a better option than deleting keys from the object.

File: registry/aws-sqs/src/index.test.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    const inputs = {}
+
+    const outputs = await sqsComponent.remove(inputs, sqsContextMock)
+
+    expect(AWS.SQS).toHaveBeenCalledTimes(1)
+    expect(AWS.mocks.deleteQueueMock).toHaveBeenCalledTimes(1)
+    expect(outputs.queueUrl).toEqual(null)
+  })
```
Comment (laardee) [inline registry/aws-sqs/src/index.test.js]: Also, add one test case which tests errors that are not omitted.

```js
it('should throw an error if something unexpected happens on delete', async () => {
    const sqsContextMock = {
      state: {
        queueName: 'myQueueOld',
        queueUrl: 'https://sqs.us-east-1.amazonaws.com/123/1234MyQueue'
      },
      archive: {},
      log: () => {},
      saveState: () => {}
    }

    AWS.mocks.deleteQueueMock = jest
      .fn()
      .mockImplementationOnce(() => Promise.reject(new Error('Something went wrong')))

    const inputs = {}

    await expect(sqsComponent.remove(inputs, sqsContextMock)).rejects.toThrow('Error: Something went wrong')

    expect(AWS.SQS).toHaveBeenCalledTimes(1)
    expect(AWS.mocks.deleteQueueMock).toHaveBeenCalledTimes(1)
  })
```

File: registry/aws-sqs/src/index.test.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    const sqsContextMock = {
+      state: {
+        queueName: 'myQueueOld',
+        queueUrl: 'https://sqs.us-east-1.amazonaws.com/123/1234MyQueue'
+      },
+      archive: {},
+      log: () => {},
+      saveState: () => {}
+    }
+
```
Comment (laardee) [inline registry/aws-sqs/src/index.test.js]: Add one time mock that throws a non-existent queue error

```js
    AWS.mocks.deleteQueueMock = jest
      .fn()
      .mockImplementationOnce(() => Promise.reject(new Error('The specified queue does not exist')))
```

This way you can test that the try-catch really works

**Comment by laardee** [comment]
Add this also to the component listings (top and bottom of the page) in the [readme](https://github.com/serverless/components/blob/master/README.md)

File: registry/aws-sqs/src/index.js (hunk 5 above)
Comment (dilantha111) [reply registry/aws-sqs/src/index.js]: Thank you very much for pointing this out. changed it.

File: registry/aws-sqs/src/index.js (hunk 6 above)
Comment (dilantha111) [reply registry/aws-sqs/src/index.js]: Agree. I think this is the best way.

**Comment by dilantha111** [comment]
Hi @laardee  I added the changes. Thank you very much again for the guidance. Btw Seems like I dismissed @brianneisler 's review by mistake ?

**Comment by brianneisler** [comment]
> Hi @laardee I added the changes. Thank you very much again for the guidance. Btw Seems like I dismissed @brianneisler 's review by mistake ?

@dilantha111 that's ok. We only require one approving review for merge. @laardee's should be enough

**Comment by dilantha111** [comment]
Hi @brianneisler  thank you. But I am not sure whether this is caused by last commit f7889dd. And build is failed at npm install stage as it seems. Any insights to help fix that.

**Comment by brianneisler** [comment]
Job restarted, let’s see if it gets stuck again

**Comment by dilantha111** [comment]
Hi @laardee  it seems like adding to the root README was causing the build fail. Is there any configurations or auto generating for that README as well ?

**Pull request merged**